distributed/protocol/tests/test_collection.py

```diff
@@ -0,0 +1,59 @@
+import pytest
+from distributed.protocol import serialize, deserialize
+import pandas as pd
+import numpy as np
+from dask.dataframe.utils import assert_eq
+
+
+@pytest.mark.parametrize("collection", [tuple, dict])
+@pytest.mark.parametrize("y,y_serializer", [(np.arange(50), "dask"), (None, "pickle")])
+def test_serialize_numpy_numpy(collection, y, y_serializer):
+    x = np.arange(100)
+    if issubclass(collection, dict):
+        header, frames = serialize({"x": x, "y": y}, serializers=("dask", "pickle"))
+    else:
+        header, frames = serialize((x, y), serializers=("dask", "pickle"))
+    t = deserialize(header, frames, deserializers=("dask", "pickle", "error"))
+
+    assert header["is-collection"] is True
+    sub_headers = header["sub-headers"]
+    assert sub_headers[0]["serializer"] == "dask"
+    assert sub_headers[1]["serializer"] == y_serializer
+    assert isinstance(t, collection)
+
+    assert ((t["x"] if isinstance(t, dict) else t[0]) == x).all()
+    if y is None:
+        assert (t["y"] if isinstance(t, dict) else t[1]) is None
+    else:
+        assert ((t["y"] if isinstance(t, dict) else t[1]) == y).all()
+
+
+@pytest.mark.parametrize("collection", [tuple, dict])
+@pytest.mark.parametrize(
+    "df2,df2_serializer",
+    [
+        (pd.DataFrame({"C": ["a", "b", None], "D": [2.5, 3.5, 4.5]}), "pickle"),
+        (None, "pickle"),
+    ],
+)
+def test_serialize_pandas_pandas(collection, df2, df2_serializer):
+    df1 = pd.DataFrame({"A": [1, 2, None], "B": [1.0, 2.0, None]})
+    if issubclass(collection, dict):
+        header, frames = serialize(
+            {"df1": df1, "df2": df2}, serializers=("dask", "pickle")
+        )
+    else:
+        header, frames = serialize((df1, df2), serializers=("dask", "pickle"))
+    t = deserialize(header, frames, deserializers=("dask", "pickle"))
+
+    assert header["is-collection"] is True
+    sub_headers = header["sub-headers"]
+    assert sub_headers[0]["serializer"] == "pickle"
+    assert sub_headers[1]["serializer"] == "pickle"
+    assert isinstance(t, collection)
+
+    assert_eq(t["df1"] if isinstance(t, dict) else t[0], df1)
+    if df2 is None:
+        assert (t["df2"] if isinstance(t, dict) else t[1]) is None
+    else:
+        assert_eq(t["df2"] if isinstance(t, dict) else t[1], df2)
```

Comment:
Here is a bad type test

```python
def test_dont_pickle_types():

    class EvilTuple(tuple):
        def __getstate__(self):
            return None

        def __setstate__(self, state):
            raise Exception("Evil Exception")

    with pytest.raises(TypeError) as info:
        serialize(EvilTuple(1, 2, 3), serializers=["dask", "msgpack"], on_error="raise")  # note, no pickle serializer

    assert "EvilTuple" in str(info.value)
```

Reply:
I've added the first two tests, but I'm not sure I understand the last one. `msgpack` doesn't seem to rely on `__{get,set}state__`, so the exception is never raised. When do we expect that exception to be raised exactly, during `msgpack.dumps(EvilTuple((1, 2, 3)))`?

Reply:
Ah, you're right, I should have made the type evil, not the object.

I tried again with the following, but it didn't fail either

```python
class Evil(type):
    def __getstate__(self):
        return None
    def __setstate__(self, state):
        raise Exception("Evil Exception")

EvilTuple = Evil("EvilTuple", (tuple,), {})

obj = EvilTuple((1, 2, 3))
pickle.loads(pickle.dumps(obj))
```

@jcrist tends to know these things better than I do.  Maybe he can suggest a failing case here?

Reply:
I think serialization with pickle is the one that should fail, no? Maybe what the test we should add is the following:

```python
def test_dont_pickle_types():

    class EvilTuple(tuple):
        def __getstate__(self):
            raise Exception("Evil Exception")

    evil = EvilTuple((1, 2, 3))

    with pytest.raises(TypeError) as info:
        serialize(evil, serializers=["pickle"], on_error="raise")

    assert "EvilTuple" in str(info.value)

    header, frames = serialize(evil, serializers=["msgpack"], on_error="raise")  # note: no pickle serializer
    assert header["serializer"] == "msgpack"
    assert deserialize(header, frames) == evil
```

The test above passes, does that cover what you wanted to check @mrocklin?

Note also that `pickle.dumps` calls `__getstate__`, not `__setstate__`.

Reply:
The attack here is that dumps succeeds (the client can successfully send the data) but that the loads errs (upon deserialization on the other end, bad things happen)

I realize now that in my test above I should have been serializing and deserializing the type, not the data, something like this:

```python
class Evil(type):
    def __getstate__(self):
        return None
    def __setstate__(self, state):
        raise Exception("Evil Exception")

EvilTuple = Evil("EvilTuple", (tuple,), {})

obj = EvilTuple((1, 2, 3))
pickle.loads(pickle.dumps(type(obj)))
```

But that still doesn't raise the exception.

Reply:
Pickling of types is hardcoded (the same way it is for functions). It looks at the qualname, and tries to import it. To make an unpickleable type just wrap the type creation in a closure:

```python
In [56]: def make_class():
    ...:     class EvilTuple(tuple):
    ...:         pass
    ...:     return EvilTuple
    ...:

In [57]: EvilTuple = make_class()

In [58]: EvilTuple
Out[58]: __main__.make_class.<locals>.EvilTuple

In [59]: import pickle

In [60]: pickle.dumps(EvilTuple)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-60-f2472b1154ee> in <module>
----> 1 pickle.dumps(EvilTuple)

AttributeError: Can't pickle local object 'make_class.<locals>.EvilTuple'
```

Reply:
Thanks @jcrist .  We're looking for something that fails on `loads`, such as would be used for nefarious behavior.  Any suggestions?

Reply:
If pickle is used for *object* serialization (not the type), then defining `__loadstate__` or a reducer that errors should work fine:

```pickle
In [65]: class ErrorOnLoad(tuple):
    ...:     def __reduce__(self):
    ...:         return ErrorOnLoad._raise, ()
    ...:     @staticmethod
    ...:     def _raise():
    ...:         raise ValueError("Oh no!")
    ...:

In [66]: e = ErrorOnLoad((1, 2, 3))

In [67]: s = pickle.dumps(e)

In [68]: pickle.loads(s)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-68-764e4625bc41> in <module>
----> 1 pickle.loads(s)

<ipython-input-65-c51f58a083ef> in _raise()
      4     @staticmethod
      5     def _raise():
----> 6         raise ValueError("Oh no!")
      7

ValueError: Oh no!
```

Reply:
We were actually concerned about deserializing the *type* (sorry to be unspecific above)

This isn't that big of a deal though, we've stopped calling `pickle.loads` in this change, so testing this here is just to be particularly safe.  It's not a hard requirement though.  That being said, thinking adversarially like this is probably a good thing for us to do.

Reply:
> We were actually concerned about deserializing the type (sorry to be unspecific above)

Since you can't override pickle for a type, you'd have to define the type in a way that works to serialize (so at module level), serialize it, then break it by making it no longer importable (e.g. by deleting it).

Something like:

```python
In [10]: class Foo(object):
    ...:     pass
    ...:

In [11]: def test():
    ...:     global Foo
    ...:     s = pickle.dumps(Foo)
    ...:     del Foo
    ...:     pickle.loads(s)
    ...:

In [12]: test()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-12-fbd55f77ab7c> in <module>
----> 1 test()

<ipython-input-11-dca8467a1b57> in test()
      3     s = pickle.dumps(Foo)
      4     del Foo
----> 5     pickle.loads(s)
      6

AttributeError: Can't get attribute 'Foo' on <module '__main__'>
```